File: modernfinance-server/Sources/App/Controllers/AnalysisController.swift
import Vapor
import Redis

struct AnalysisController {
    
    // Get base analysis template for a symbol
    func getBaseTemplate(req: Request) async throws -> CachedAnalysisTemplate {
        guard let symbol = req.parameters.get("symbol") else {
            throw Abort(.badRequest, reason: "Symbol is required")
        }
        
        let cacheService = req.application.make(CacheService.self)
        
        // Check cache first
        if let cached = try await cacheService.getAnalysisTemplate(symbol: symbol, on: req) {
            req.logger.info("Cache hit for symbol: \(symbol)")
            return cached
        }
        
        req.logger.info("Cache miss for symbol: \(symbol), generating template")
        
        // Generate base template
        let template = try await generateBaseTemplate(symbol: symbol, on: req)
        
        // Cache for 1 hour
        try await cacheService.setAnalysisTemplate(template, ttl: 3600, on: req)
        
        return template
    }
    
    // Cache analysis data
    func cacheAnalysisData(req: Request) async throws -> HTTPStatus {
        let cacheRequest = try req.content.decode(AnalysisCacheRequest.self)
        
        let cacheService = req.application.make(CacheService.self)
        
        switch cacheRequest.cacheLevel {
        case .base:
            // Cache only deterministic data
            let template = createTemplateFromAnalysis(cacheRequest.analysisData)
            try await cacheService.setAnalysisTemplate(template, ttl: 3600, on: req)
            
        case .partial:
            // Cache base + some agent perspectives
            let template = createTemplateFromAnalysis(cacheRequest.analysisData, includeAgents: true)
            try await cacheService.setAnalysisTemplate(template, ttl: 1800, on: req)
            
        case .full:
            // Cache complete analysis (not recommended)
            req.logger.warning("Full caching requested for \(cacheRequest.symbol) - consider using partial caching")
            try await cacheService.setFullAnalysis(cacheRequest.analysisData, ttl: 900, on: req)
        }
        
        return .ok
    }
    
    // Get cached analysis
    func getCachedAnalysis(req: Request) async throws -> ComprehensiveStockAnalysis? {
        guard let symbol = req.parameters.get("symbol") else {
            throw Abort(.badRequest, reason: "Symbol is required")
        }
        
        let cacheService = req.application.make(CacheService.self)
        return try await cacheService.getFullAnalysis(symbol: symbol, on: req)
    }
    
    // MARK: - Private Methods
    
    private func generateBaseTemplate(symbol: String, on req: Request) async throws -> CachedAnalysisTemplate {
        // This would integrate with your existing data sources
        // For now, returning a mock template
        
        let marketDataService = req.application.make(MarketDataService.self)
        let fundamentals = try await marketDataService.getFundamentals(symbol: symbol, on: req)
        let sentiment = try await marketDataService.getSentiment(symbol: symbol, on: req)
        
        // Generate competitive position using AI
        let competitivePosition = CompetitiveAnalysis(
            marketPosition: "Market Leader",
            competitiveAdvantages: ["Strong brand", "Network effects", "High switching costs"],
            threatsFromCompetitors: ["New entrants", "Tech disruption"],
            barrierToEntry: 8.5,
            switchingCosts: 7.0,
            networkEffects: 9.0,
            brandStrength: 8.0
        )
        
        // Generate base agent perspectives (deterministic parts)
        let baseAgentPerspectives = generateBaseAgentPerspectives(fundamentals: fundamentals)
        
        return CachedAnalysisTemplate(
            symbol: symbol,
            fundamentals: fundamentals,
            marketSentiment: sentiment,
            competitivePosition: competitivePosition,
            baseAgentPerspectives: baseAgentPerspectives,
            cacheTimestamp: Date(),
            ttl: 3600
        )
    }
    
    private func createTemplateFromAnalysis(_ analysis: ComprehensiveStockAnalysis, includeAgents: Bool = false) -> CachedAnalysisTemplate {
        return CachedAnalysisTemplate(
            symbol: analysis.symbol,
            fundamentals: analysis.fundamentalMetrics,
            marketSentiment: MarketSentiment(
                analystRating: 4.0,
                socialSentiment: 0.7,
                newsVolume: 150,
                sentimentTrend: "positive"
            ),
            competitivePosition: analysis.swarmDCF?.competitivePosition ?? CompetitiveAnalysis(
                marketPosition: "Unknown",
                competitiveAdvantages: [],
                threatsFromCompetitors: [],
                barrierToEntry: 5.0,
                switchingCosts: 5.0,
                networkEffects: 5.0,
                brandStrength: 5.0
            ),
            baseAgentPerspectives: includeAgents ? analysis.agentPerspectives : [],
            cacheTimestamp: Date(),
            ttl: 3600
        )
    }
    
    private func generateBaseAgentPerspectives(fundamentals: [String: Double]) -> [AgentPerspective] {
        // Generate deterministic base perspectives based on fundamentals
        let peRatio = fundamentals["P/E Ratio"] ?? 20.0
        let revenueGrowth = fundamentals["Revenue Growth"] ?? 0.0
        
        return [
            AgentPerspective(
                agentType: .fundamental,
                recommendation: peRatio < 15 ? "Buy" : peRatio > 30 ? "Sell" : "Hold",
                targetPrice: 0.0, // Will be calculated on device
                reasoning: "Based on fundamental analysis",
                confidence: 0.8,
                keyPoints: ["P/E ratio analysis", "Revenue growth trends"],
                bias: .neutral
            ),
            AgentPerspective(
                agentType: .technical,
                recommendation: "Hold",
                targetPrice: 0.0,
                reasoning: "Technical indicators suggest consolidation",
                confidence: 0.7,
                keyPoints: ["RSI neutral", "Moving averages converging"],
                bias: .neutral
            )
        ]
    }
}

File: VERSION-2_modernfinance_server/src/controllers/analysis.controller.ts
import { Request, Response } from 'express';
import { analysisService, CacheLevel } from '../services/analysis.service';
import { logger } from '../utils/logger';

export class AnalysisController {
  async getBaseTemplate(req: Request, res: Response) {
    try {
      const { symbol } = req.params;
      
      if (!symbol) {
        return res.status(400).json({ error: 'Symbol is required' });
      }

      // Check cache first
      const cached = await analysisService.getCachedTemplate(symbol);
      if (cached) {
        res.set('X-Cache-Hit', 'true');
        res.set('Cache-Control', 'public, max-age=3600');
        return res.json(cached);
      }

      // Generate new template with real data and AI perspectives
      const template = await analysisService.generateBaseTemplate(symbol);
      
      // Cache the complete template including agent perspectives
      await analysisService.cacheTemplate(symbol, template);

      res.set('X-Cache-Hit', 'false');
      res.set('Cache-Control', 'public, max-age=3600');
      res.json(template);
    } catch (error) {
      logger.error('Error getting base template:', error);
      res.status(500).json({ 
        error: 'Failed to generate base template',
        message: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }

  async cacheAnalysisData(req: Request, res: Response) {
    try {
      const { symbol, analysisData, cacheLevel } = req.body;
      
      if (!symbol || !analysisData) {
        return res.status(400).json({ error: 'Symbol and analysisData are required' });
      }

      // Determine cache level from request or data
      const level = cacheLevel || (analysisData.cacheLevel === 'full' ? CacheLevel.FULL :
                                   analysisData.cacheLevel === 'partial' ? CacheLevel.PARTIAL :
                                   CacheLevel.BASE);

      // Cache the complete analysis data including agent perspectives
      await analysisService.cacheCompleteAnalysis(symbol, analysisData, level);
      
      res.json({ 
        success: true, 
        message: 'Analysis cached successfully',
        cacheLevel: level
      });
    } catch (error) {
      logger.error('Error caching analysis data:', error);
      res.status(500).json({ 
        error: 'Failed to cache analysis data',
        message: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }

  async getCachedAnalysis(req: Request, res: Response) {
    try {
      const { symbol } = req.params;
      
      if (!symbol) {
        return res.status(400).json({ error: 'Symbol is required' });
      }

      const cached = await analysisService.getCachedAnalysis(symbol);
      
      if (!cached) {
        return res.status(404).json({ error: 'No cached analysis found for symbol' });
      }

      // Set appropriate cache headers
      res.set('Cache-Control', 'public, max-age=3600');
      res.set('X-Cache-Hit', 'true');
      
      res.json(cached);
    } catch (error) {
      logger.error('Error getting cached analysis:', error);
      res.status(500).json({ 
        error: 'Failed to retrieve cached analysis',
        message: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }

  async getCachedAgentPerspectives(req: Request, res: Response) {
    try {
      const { symbol } = req.params;
      
      if (!symbol) {
        return res.status(400).json({ error: 'Symbol is required' });
      }

      const perspectives = await analysisService.getCachedAgentPerspectives(symbol);
      
      if (!perspectives || perspectives.length === 0) {
        // Return empty array instead of 404 to match client expectations
        res.set('X-Cache-Hit', 'false');
        return res.json([]);
      }

      res.set('X-Cache-Hit', 'true');
      res.set('Cache-Control', 'public, max-age=1800');
      res.json(perspectives);
    } catch (error) {
      logger.error('Error getting cached agent perspectives:', error);
      // Return empty array on error to match client expectations
      res.json([]);
    }
  }
}

export const analysisController = new AnalysisController();